File: may/aho.py
mydic={}
MAXS=1000
MAXC=26
out=[0]*MAXC
f=[0]*MAXC
g=[[0]*MAXC]*MAXS
alph={}
# # def init():
#     for i in range('a',27):
#         print(i)
        
        
def build_matching_machine():
    global f
    global out
    global g
    global mydic

    for i in range(len(out)):
        out[i]=0
        
    g=[[-1]*26]*1000
        
    states=1
    counter=0
    for i in mydic:
        curr_state=0
        for j in range(len(i)):
            ch=ord(i[j])-ord('a')
            if g[curr_state][ch] == -1:
                g[curr_state][ch] = states
                states=states+1
            curr_state=g[curr_state][ch]
            
        out[curr_state] = out[curr_state] | (1 << counter)
    
    for i in f:
        i=-1

    for ch in range(26):
        if g[0][ch] == -1:
            g[0][ch] = 0
    
    q=[]
    for ch in range(26):
        if g[0][ch] != 0:
            f[g[0][ch]] = 0
            q.append(g[0][ch])
    print(q)
    while len(q)!=0:
        state=q[0]
        q.pop(0)
        for ch in range(27):
            print(q)
            if g[state][ch]!=-1:
                failure=f[state]
                while g[failure][ch] ==-1:
                    failure=f[failure]
                
                failure=g[failure][ch]
                f[g[state][ch]]=failure
                out[g[state][ch]]=out[g[state][ch]] | out[failure]
                q.append(g[state][ch])
                try:
                    q.remove(0)
                except:
                    print(q)
                
    return states

def findnextstae(current_state,next_input):
    global g
    global f
    answer=current_state
    ch=ord(next_input)-ord('a')
    while g[answer][ch] == -1:
        answer = f[answer]
    
    return g[answer][ch]

def search_word(text):
    global f
    global out
    global g
    global mydic
    ans=0
    k=mydic.size()
    curr_state=0
    for i in range(len(text)):
        curr_state=findnextstae(curr_state,text[i])
        if out[curr_state] == 0:
            continue
        counter=0
        for j in mydic:
            if out[curr_state] & (1<<counter):
                ans = ans + mydic[j]
                counter = counter + 1                
    return ans

    
            
        
    
for _ in range(int(input())):
    a=input()
    b=input()
    # init()
    q=int(input())
    mydic.clear()
    for i in range(q):
        temp,val=input().split()
        intval=int(val)
        mydic[temp]=intval
        
    print(mydic)
    build_matching_machine()

File: may/aho.cpp
#include <bits/stdc++.h>
#define ios ios_base::sync_with_stdio(false);cin.tie(NULL)
#define mod 1000000007
#define INF 1000000000
#define endl '\n'
#define cout(x) cout<<x<<'\n'
#define debug(x) cout<<x<<'\n\n'
#define cin(x) cin>>x
#define cin2(x,y) cin>>x>>y
#define cin3(x,y,z) cin>>x>>y>>z


using namespace std;
const int MAXS = 1000; 

// Maximum number of characters in input alphabet 
const int MAXC = 26; 
int out[MAXS]; 
unordered_map<string,int> mpq;
int f[MAXS];
int g[MAXS][MAXC];
int buildMatchingMachine() 
{ 
	int k=mpq.size();
	// Initialize all values in output function as 0. 
	memset(out, 0, sizeof out); 

	// Initialize all values in goto function as -1. 
	memset(g, -1, sizeof g); 

	// Initially, we just have the 0 state 
	int states = 1; 
	int counter=0;
	for (auto i=mpq.begin(); i!=mpq.end(); ++i,counter++) 
	{ 
		const string &word = i->first; 
		int currentState = 0; 

		// Insert all characters of current word in arr[] 
		for (int j = 0; j < word.size(); ++j) 
		{ 
			int ch = word[j] - 'a';  
			if (g[currentState][ch] == -1) 
				g[currentState][ch] = states++; 

			currentState = g[currentState][ch]; 
		} 

		// Add current word in output function 
		out[currentState] |= (1 << counter); 
	} 

	// For all characters which don't have an edge from 
	// root (or state 0) in Trie, add a goto edge to state 
	// 0 itself 
	for (int ch = 0; ch < MAXC; ++ch) 
		if (g[0][ch] == -1) 
			g[0][ch] = 0; 
	memset(f, -1, sizeof f); 

	// Failure function is computed in breadth first order 
	// using a queue 
	queue<int> q; 
	for (int ch = 0; ch < MAXC; ++ch){
		if (g[0][ch] != 0) 
		{ 
			f[g[0][ch]] = 0; 
			q.push(g[0][ch]); 
		} 
	} 

	// Now queue has states 1 and 3 
	while (q.size()){ 
		int state = q.front(); 
		q.pop(); 
		for (int ch = 0; ch <= MAXC; ++ch){
			if (g[state][ch] != -1){ 
				int failure = f[state]; 
				while (g[failure][ch] == -1) 
					failure = f[failure]; 
				failure = g[failure][ch]; 
				f[g[state][ch]] = failure; 
				out[g[state][ch]] |= out[failure];  
				q.push(g[state][ch]); 
			} 
		} 
	} 
	return states; 
}
int findNextState(int currentState, char nextInput){ 
	int answer = currentState; 
	int ch = nextInput - 'a'; 
	while (g[answer][ch] == -1) 
		answer = f[answer];
	return g[answer][ch]; 
}
int searchWords(string text){ 
    int ans=0;
	int k=mpq.size();
	int currentState = 0; 
	for (int i = 0; i < text.size(); ++i){ 
		currentState = findNextState(currentState, text[i]);  
		if (out[currentState] == 0) 
			continue; 
		int counter=0;
		for (auto j = mpq.begin(); j!=mpq.end(); ++j,++counter){ 
			if(out[currentState]&(1<<counter)){ 
				ans+=mpq[j->first];
			} 
		} 
	} 
	return ans;
}
int32_t main(){
    ios;
    int t;
    cin>>t;
    while(t--){
        string a,b;
        cin2(a,b);
        int q;
        cin(q);
        for(int i=0;i<q;i++){
            string temp;
			int val;
			cin2(temp,val);
            mpq[temp]=val;
        }
		buildMatchingMachine(); 
		int max_pleasue=INT_MIN;
		string left="",right="";
        for(int i=0;i<a.size();i++){
			left=left+a[i];
			for(int j=b.size()-1;j>=0;j--){
				right=b[j]+right;
				cout<<left+right<<endl;
				max_pleasue=max(max_pleasue,searchWords(left+right));
			}
			right="";
		}
        cout(max_pleasue);
		mpq.clear();
    }
    return 0;
}